neonvm/controllers/vm_controller_unit_test.go

```diff
@@ -0,0 +1,262 @@
+	m.Called(object, annotations, eventtype, reason, messageFmt, args)
+}
+
+// defaultVm returns a VM which is similar to what we can reasonably
+// expect from the control plane.
+func defaultVm() *vmv1.VirtualMachine {
+	return &vmv1.VirtualMachine{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test-vm",
+			Namespace: "default",
+		},
+		Spec: vmv1.VirtualMachineSpec{
+			EnableSSH:          lo.ToPtr(false),
+			EnableAcceleration: lo.ToPtr(true),
+			//nolint:exhaustruct // This is a test
+			Guest: vmv1.Guest{
+				KernelImage:         lo.ToPtr("kernel-img"),
+				AppendKernelCmdline: nil,
+				CPUs: vmv1.CPUs{
+					Min: lo.ToPtr(vmv1.MilliCPU(1000)),
+					Max: lo.ToPtr(vmv1.MilliCPU(2000)),
+					Use: lo.ToPtr(vmv1.MilliCPU(1500)),
+				},
+				MemorySlots: vmv1.MemorySlots{
+					Min: lo.ToPtr(int32(512)),
+					Max: lo.ToPtr(int32(2048)),
+					Use: lo.ToPtr(int32(1024)),
+				},
+				MemorySlotSize: *resource.NewQuantity(1, resource.DecimalSI),
+			},
+		},
+		Status: vmv1.VirtualMachineStatus{
+			Phase:         "",
+			Conditions:    []metav1.Condition{},
+			RestartCount:  0,
+			PodName:       "",
+			PodIP:         "",
+			ExtraNetIP:    "",
+			ExtraNetMask:  "",
+			Node:          "",
+			CPUs:          lo.ToPtr(vmv1.MilliCPU(100)),
+			MemorySize:    resource.NewQuantity(123, resource.DecimalSI),
+			SSHSecretName: "",
+		},
+	}
+}
+
+type testParams struct {
+	t            *testing.T
+	ctx          context.Context
+	r            *VMReconciler
+	client       client.Client
+	origVM       *vmv1.VirtualMachine
+	mockRecorder *mockRecorder
+}
+
+var reconcilerMetrics = MakeReconcilerMetrics()
+
+func newTestParams(t *testing.T) *testParams {
+	os.Setenv("VM_RUNNER_IMAGE", "vm-runner-img")
+
+	logger := zap.New(zap.UseDevMode(true), zap.WriteTo(os.Stdout),
+		zap.Level(zapcore.DebugLevel))
+	ctx := log.IntoContext(context.Background(), logger)
+
+	scheme := runtime.NewScheme()
+	scheme.AddKnownTypes(vmv1.SchemeGroupVersion, &vmv1.VirtualMachine{})
+	scheme.AddKnownTypes(corev1.SchemeGroupVersion, &corev1.Pod{})
+
+	params := &testParams{
+		t:      t,
+		ctx:    ctx,
+		client: fake.NewClientBuilder().WithScheme(scheme).Build(),
+		//nolint:exhaustruct // This is a mock
+		mockRecorder: &mockRecorder{},
+		r:            nil,
+		origVM:       nil,
+	}
+
+	params.r = &VMReconciler{
+		Client:   params.client,
+		Recorder: params.mockRecorder,
+		Scheme:   scheme,
+		Config: &ReconcilerConfig{
+			IsK3s:                   false,
+			UseContainerMgr:         false,
+			MaxConcurrentReconciles: 10,
+			QEMUDiskCacheSettings:   "",
+			FailurePendingPeriod:    time.Minute,
+			FailingRefreshInterval:  time.Minute,
+		},
+		Metrics: reconcilerMetrics,
+	}
+
+	return params
+}
+
+func (p *testParams) initVM(vm *vmv1.VirtualMachine) *vmv1.VirtualMachine {
+	err := p.client.Create(p.ctx, vm)
+	require.NoError(p.t, err)
+	p.origVM = vm
+
+	// Do serialize/deserialize, to normalize resource.Quantity
+	return p.getVM()
```

Comment:
out of curiosity, does this actually work? (asking because it has somewhat surprising (IMO) implications about the mock client)

Reply:
Huh! TIL :)

Yep, makes sense.